File: February2022/Java/RemoveDuplicatesFromSortedArrayII.java
// Author - @saorav21994
// TC - O(n)



class Solution {
    public int removeDuplicates(int[] nums) {
        int l = nums.length;
        int prev = nums[0];
        int cnt = 1;    // maximum duplicates allowed track cnt (= 2)
        int pidx = 1;   // index on which element to be replaced if needed
        for (int i = 1; i < l; i++) {
            if (prev == nums[i]) {
                if (cnt < 2) {
                    cnt += 1;
                    nums[pidx] = nums[i];
                    pidx += 1;
                }
                else {      // cnt >= 2, no more duplicate allowed
                    // pass
                }
            }
            else {
                prev = nums[i];
                cnt = 1;
                nums[pidx] = nums[i];
                pidx += 1;
            }
        }
        return pidx;
    }
}

File: February2022/Kotlin/Remove_Duplicates_from_Sorted_Array_II.kt
// Author : Shobhit Behl(shobhitbruh)
// T:O(N) 
class Solution {
    fun removeDuplicates(nums: IntArray): Int {
        var ans = 0
        var j = 0
        var f = 0
        var l = 0
        for (i in nums.indices) {
            if (nums[i] == l && f == 2) {
                continue
            } else {
                if (nums[i] != l) f = 0
                nums[j] = nums[i]
                f++
                l = nums[i]
                ans++
                j++
            }
        }
        return ans
    }
}